test/cqlpy/test_collection_subscript.py

```diff
@@ -0,0 +1,150 @@
+    assert list(cql.execute(lookup1, [2, p])) == [(20,)]
+    assert list(cql.execute(lookup1, [3, p])) == [(None,)]
+
+    lookup2 = cql.prepare(f"SELECT m[:x1], m[:x2] FROM {table1} WHERE p = :key")
+    assert list(cql.execute(lookup2, {'x1':2, 'x2':1, 'key':p})) == [(20,10)]
+
+def test_null_map(cql, table1):
+    p = unique_key_int()
+    cql.execute(f"INSERT INTO {table1}(p) VALUES ({p})")
+    assert list(cql.execute(f"SELECT m[1] FROM {table1} WHERE p={p}")) == [(None,)]
+
+# scylla only because scylla returns null while cassandra returns error
+def test_null_subscript(scylla_only, cql, table1):
+    p = unique_key_int()
+    cql.execute(f"INSERT INTO {table1}(p,m) VALUES ({p}, " + "{1:10,2:20})")
+    assert list(cql.execute(f"SELECT m[null] FROM {table1} WHERE p={p}")) == [(None,)]
+
+def test_subscript_and_field(cql, test_keyspace):
+    with new_type(cql, test_keyspace, '(a int)') as typ:
+        schema = f"p int PRIMARY KEY, m map<int, frozen<{typ}>>"
+        with new_test_table(cql, test_keyspace, schema) as table:
+            p = unique_key_int()
+            cql.execute(f"INSERT INTO {table}(p,m) VALUES ({p}, " + "{1:{a:10}})")
+            assert list(cql.execute(f"SELECT m[1].a FROM {table} WHERE p={p}")) == [(10,)]
+
+def test_field_and_subscript(cql, test_keyspace):
+    with new_type(cql, test_keyspace, '(a frozen<map<int,int>>)') as typ:
+        schema = f"p int PRIMARY KEY, t {typ}"
+        with new_test_table(cql, test_keyspace, schema) as table:
+            p = unique_key_int()
+            cql.execute(f"INSERT INTO {table}(p,t) VALUES ({p}, " + "{a:{1:10}})")
+            assert list(cql.execute(f"SELECT t.a[1] FROM {table} WHERE p={p}")) == [(10,)]
+
+def test_field_and_subscript_and_field(cql, test_keyspace):
+    with new_type(cql, test_keyspace, '(b int)') as typ1, \
+         new_type(cql, test_keyspace, f"(a frozen<map<int,{typ1}>>)") as typ2:
+            schema = f"p int PRIMARY KEY, t {typ2}"
+            with new_test_table(cql, test_keyspace, schema) as table:
+                p = unique_key_int()
+                cql.execute(f"INSERT INTO {table}(p,t) VALUES ({p}, " + "{a:{1:{b:10}}})")
+                assert list(cql.execute(f"SELECT t.a[1].b FROM {table} WHERE p={p}")) == [(10,)]
+
+def test_other_types_cannot_be_subscripted(cql, table1):
+    with pytest.raises(InvalidRequest, match='not a'):
+        cql.execute(f"SELECT p[2] FROM {table1}")
+    with pytest.raises(InvalidRequest, match='not a'):
+        cql.execute(f"SELECT token(p)[2] FROM {table1}")
+
+def test_udf_subscript(scylla_only, cql, test_keyspace, table1):
+    fn = "(k int) CALLED ON NULL INPUT RETURNS int LANGUAGE Lua AS 'return k+1'"
+    with new_function(cql, test_keyspace, fn, 'add_one'):
+        p = unique_key_int()
+        cql.execute(f"INSERT INTO {table1}(p,m) VALUES ({p}, " + "{1:10,2:20})")
+        assert list(cql.execute(f"SELECT m[add_one(1)] FROM {table1} WHERE p={p}")) == [(20,)]
```

Comment:
Please also check IF clause with UDF.

Also test that we reject aggregate functions for IF, since this patch brings in UDF support for IF. 

We do:

```
cql3/statements/modification_statement.cc:    expr::verify_no_aggregate_functions(prepared, "IF clause");
```

but let's have a test.

Reply:
I tested udf in a condition: `UPDATE t SET m[1]=11 WHERE p=0 IF m[add_one(1)]=10`
and it fails with `User function cannot be executed in this context`
but I think this syntax is valid also before the PR and the issue is existing, no?

Reply:
It's acceptable. Either UDF being accepted and working would be good, or UDF being outright rejected. What wouldn't be good is UDF being accepted and then crashing due to not having the infrastructure to provide the UDF with a thread.

It's not a bug, it's a limitation of the current code. We reject many expressions that the grammar accepts.

We should work to reduce those limitations, but not here.

Reply:
I think there are cases where a UDF is crashing but it's an already existing issue. For example https://github.com/scylladb/scylladb/issues/13746
There is one problematic case I noticed related to this PR which is `SELECT udf(m[1])` and I rejected it with an error.

Reply:
Why does it crash? It should work.

Reply:
What assignment?

Reply:
I mean in `SELECT udf(m[1])` it calls `test_assignment` which is currently not implemented for a subscript expression

Reply:
Ok.